File: v2/src/main/java/com/api/v2/people/services/PersonRegistrationServiceImpl.java
package com.api.v2.people.services;

import com.api.v2.people.domain.exposed.Person;
import com.api.v2.people.domain.PersonRepository;
import com.api.v2.people.requests.PersonRegistrationDto;
import com.api.v2.people.services.exposed.PersonRegistrationService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
@RequiredArgsConstructor
public abstract class PersonRegistrationServiceImpl implements PersonRegistrationService {

    private final PersonRepository personRepository;

    @Override
    public Mono<Person> register(@Valid PersonRegistrationDto registrationDto) {
        return personRepository
                .find(registrationDto.ssn(), registrationDto.email())
                .singleOptional()
                .flatMap(optional -> {
                    if (optional.isPresent()) {
                        return Mono.just(optional.get());
                    }
                    Person person = Person.of(registrationDto);
                    return personRepository.save(person);
                });
    }
}

File: v2/src/main/kotlin/com/api/v2/people/services/PersonRegistrationServiceImpl.kt
package com.api.v2.people.services

import com.api.v2.people.Person
import com.api.v2.people.PersonRegistrationService
import com.api.v2.people.domain.PersonCrudRepository
import com.api.v2.people.requests.PersonRegistrationDTO
import jakarta.validation.Valid
import org.springframework.stereotype.Service

@Service
class PersonRegistrationServiceImpl(val personCrudRepository: PersonCrudRepository): PersonRegistrationService {

    override fun register(registrationDTO: @Valid PersonRegistrationDTO): Person {
        val foundPerson = personCrudRepository.findBySINOrEmail(registrationDTO.sin, registrationDTO.email)
        if (foundPerson == null) {
            val newPerson = Person.of(registrationDTO)
            return personCrudRepository.save(newPerson)
        }
        return foundPerson
    }
}